docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+
+The OpenVPN server running in the shoot cluster is configured as follows:
+
+```
+mode server
+tls-server
+proto tcp4-server
+dev tun0
+
+[...]
+
+server 192.168.123.0 255.255.255.0
+
+push "route 10.243.0.0 255.255.128.0"
+push "route 10.243.128.0 255.255.128.0"
+
+duplicate-cn
+
+key "/srv/secrets/vpn-shoot/tls.key"
+cert "/srv/secrets/vpn-shoot/tls.crt"
+ca "/srv/secrets/vpn-shoot/ca.crt"
+dh "/srv/secrets/dh/dh2048.pem"
+
+tls-auth "/srv/secrets/tlsauth/vpn.tlsauth" 0
+push "route 10.242.0.0 255.255.0.0"
+```
+
+It is a TCP TLS server and configured to automatically assign IP addresses for OpenVPN clients (`server` directive). In addition, it pushes the shoot cluster node-, pod-, and service ranges to the clients running in the seed cluster (`push` directive). 
+
+**Note:** The network mesh spanned by OpenVPN uses the network range `192.168.123.0 - 192.168.123.255`. This network range cannot be used in either shoot-, or seed clusters. If it is used this might cause subtle problem due to network range overlaps. Unfortunately, this appears not to be well documented but this restriction exists since the very beginning.
+
+# Proposal
+
+The general idea of the proposal is to keep the existing cluster vpn solution more or less as is, but change the underlying network used for the `vpn seed => vpn shoot` connection. The underlying network should be established in the reversed direction, i.e. the shoot cluster should initiate the network connection, but it nevertheless should work in both directions.
+
+We achieve this by tunneling the open vpn connection through a WireGuard tunnel, which is established from the shoot to the seed (note that WireGuard uses udp as protocol). Independent of that we can also use udp for the open vpn connection, but we can also stay with tcp as it was before. While this might look like a big change, it only introduces minor changes to the existing solution, but let's look at the details. In essence, the OpenVPN connection does not require a public endpoint in the shoot cluster but it used the internal endpoint provided by the Wireguard tunnel.
+
+The WireGuard network needs a separate network range/cidr. It has to be unique for the seed and all its shoot clusters. An example for an assumed workload of around 250 shoot clusters with enough buffer would be 192.168.0.0/22 (1024 IP addresses), i.e. 192.168.0.0-192.168.3.255. The IP addresses from this range need to be managed, but the IP address management (IPAM) using the gardener Kubernetes objects like seed and shootstate as backing store is fairly straight forward. This is especially true as we do not expect large network ranges and only infrequent IP allocations. Hence, the IP address allocation can be quite simple, i.e. scan the range for a free IP address.
```

Comment:
I would rather plan with ~1000 shoots per seed. Today, ~250 is the default, but this might change in the future, and imposing "artificial" limits here might hit us in the long run.

Reply:
Thanks for the heads up. This means we will have to test. Let me think of something...

Reply:
While we need to gather more data concerning the performance implications with regards to wireguard, otherwise there should not be an "artificial" limit in this regard. The network range was just an example. It can be extended as required (as long as there is enough IP space available, but the 192.168.0.0/16 range seems big enough, especially as seed may have overlapping ranges).

Reply:
I have adjusted the document